pennylane/optimize/shot_adaptive.py

```diff
@@ -0,0 +1,453 @@
+    Step 1: cost = -2.983999999999999, shots_used = 336
+    Step 2: cost = -4.974, shots_used = 624
+    Step 3: cost = -5.534, shots_used = 1054
+    Step 4: cost = -6.5, shots_used = 1798
+    Step 5: cost = -6.684, shots_used = 2942
+    Step 6: cost = -6.992, shots_used = 4350
+    Step 7: cost = -6.970000000000001, shots_used = 5814
+    Step 8: cost = -6.998, shots_used = 7230
+    Step 9: cost = -6.686, shots_used = 9006
+    Step 10: cost = -6.85, shots_used = 11286
+    Step 11: cost = -6.628000000000001, shots_used = 14934
+    Step 12: cost = -6.862, shots_used = 17934
+    Step 13: cost = -7.188, shots_used = 22950
+    Step 14: cost = -6.9879999999999995, shots_used = 28302
+    Step 15: cost = -7.380000000000001, shots_used = 34134
+    Step 16: cost = -7.656, shots_used = 41022
+    Step 17: cost = -7.212, shots_used = 48918
+    Step 18: cost = -7.532, shots_used = 56286
+    Step 19: cost = -7.462000000000001, shots_used = 63822
+    Step 20: cost = -7.306000000000001, shots_used = 72534
+    Step 21: cost = -7.234, shots_used = 82014
+    Step 22: cost = -7.314, shots_used = 92838
+    """
+
+    def __init__(self, min_shots, weighted_random_sampling=True, mu=0.99, b=1e-6, stepsize=0.07):
+        self.wrs = weighted_random_sampling
+        self.trainable_args = set()
+
+        # hyperparameters
+        self.min_shots = min_shots
+        self.mu = mu  # running average constant
+        self.b = b  # regularization bias
+        self.lipschitz = None
+
+        # keep track of the total number of shots used
+        self.shots_used = 0
+        # total number of iterations
+        self.k = 0
+        # Number of shots per parameter
+        self.s = None
+        # maximum number of shots allowed per iteration
+        self.max_shots = None
+
+        # Running average of the parameter gradients
+        self.chi = None
+        # Running average of the variance of the parameter gradients
+        self.xi = None
+
+        super().__init__(stepsize=stepsize)
+
+    @staticmethod
+    def weighted_random_sampling(qnodes, coeffs, shots, argnums, *args, **kwargs):
+        """Returns an array containing length ``shots`` single-shot estimates
+        of the Hamiltonian gradient. The shots are distributed randomly over
+        the terms in the Hamiltonian, as per a multinomial distribution.
+
+        Args:
+            qnodes (Sequence[.tape.QNode]): Sequence of QNodes, each one when evaluated
+                returning the corresponding expectation value of a term in the Hamiltonian.
+            coeffs (Sequence[float]): Sequences of coefficients corresponding to
+                each term in the Hamiltonian.
+            shots (int): The number of shots used to estimate the Hamiltonian expectation
+                value. These shots are distributed over the terms in the Hamiltonian,
+                as per a Multinomial distribution.
+            argnums (Sequence[int]): the QNode argument indices which are trainable

```

Comment:
isn't this a property of the qnode?

Reply:
Yes, but allow it to be passed separately here for two reasons:

* the cost function might not be a bare QNode; it could also be an `ExpvalCost`.
* it is technically an attribute of the tape, and the tape will not exist before the QNode has been evaluated! So easier to just extract it manually.